### PR TITLE
[DNM] [WIP] [NEED SQL PEOPLE] Fixes the non-stop server crashes

### PR DESCRIPTION
**Pull request by Iamgoofball**
# Need people who know how to do SQL queries to help rewrite all the queries. Inquire in coderbus and ping me there.

## About The Pull Request

Switches the SQL stuff to be based on rust-g. Fixes the crashes.

## Why It's Good For The Game

Read the title.

## Changelog
:cl:
server: The database now uses rust-g to talk to it.
/:cl:

todo:

- [ ] Ask the big brain byond hackers how to make this properly async without crashing the server. after the first five SQL queries.
- [ ] Get all the SQL shit converted to the new system.
- [ ] Ask people why they've worked on all this other cool shit for the server but can't be bothered to fix the critical crashing issue that's plagued the servers for 6 months.

Seriously. Our host had time to complain about how github actions works, but not enough time to work on fixing this critical crash.

Come on guys. Come on. Why are the only people who seem to care about fixing this known issue me, oranges, and mrty?

### CODE REVIEW

**Comment by ATH1909** [comment]
Fulpstation seems to care, if the thing you're fixing here happened to be the cause of the abnormally high crash rate that Fulpstation currently has.

**Comment by remove32** [comment]
> 
> 
> Fulpstation seems to care, if the thing you're fixing here happened to be the cause of the abnormally high crash rate that Fulpstation currently has.

i have no idea how to check the cause of the crashes on fulp and currently don't have time to investigate. all i know is that everything points to out-of-memory exceptions. if that's related to this i'd be interested.

File: code/controllers/subsystem/dbcore.dm [hunk 1]
```diff
@@ -272,131 +235,78 @@ Delayed insert mode was removed in mysql 7 and only works with MyISAM type table
...
-	if(!async)
-		. = !query.WaitForCompletion()
+	world.log << sql
+	world.log << parameters
+	in_progress = TRUE
+	if(async)
+		query_results = rustg_sql_query_async(sql, parameters)
+		world.log << query_results
+		var/done = 0
+		while(!done)
```
Comment (optimumtact) [inline code/controllers/subsystem/dbcore.dm:284]: you know you're supposed to sleep in this loop right?

**Comment by optimumtact** [comment]
Unless you see a high number of crashes in DD with BSQL.dll like we do (which really we only encounter because some of our servers are far from the DB), I doubt this is really a major issue for your server.

**Comment by remove32** [comment]
> 
> 
> Unless you see a high number of crashes in DD with BSQL.dll like we do (which really we only encounter because some of our servers are far from the DB), I doubt this is really a major issue for your server.

not likely then considering the infrastructure i'm utilizing. thanks for the insight

File: code/controllers/subsystem/dbcore.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/dbcore.dm:284]: we're allowed to use sleep()?

File: code/controllers/subsystem/dbcore.dm [hunk 2]
```diff
@@ -62,68 +51,50 @@ SUBSYSTEM_DEF(dbcore)
...
+/datum/controller/subsystem/dbcore/proc/TestArbitraryDBCallDebugging(query, parameters = "{}")
+	world.log << "RUNNING QUERY [query]"
+	var/query_results = rustg_sql_query_blocking(query, parameters)
+	world.log << query_results
+	world.log << "QUERY OVER"
 
 /datum/controller/subsystem/dbcore/proc/CheckSchemaVersion()
 	if(CONFIG_GET(flag/sql_enabled))
-		if(Connect())
+		if(rustg_sql_connected())
```
Comment (steamp0rt) [inline code/controllers/subsystem/dbcore.dm:80]: this returns json too.
`{"status": "offline"}` or `{"status": "online"}`

worst case scenario, `{"status": "err", "data": "SomethingError: something bad happened!"}`

no wo

File: code/controllers/subsystem/dbcore.dm [hunk 3]
```diff
@@ -135,33 +106,22 @@ SUBSYSTEM_DEF(dbcore)
...
-
-/datum/controller/subsystem/dbcore/proc/IsConnected()
-	if(!CONFIG_GET(flag/sql_enabled))
-		return FALSE
-	//block until any connect operations finish
-	var/datum/BSQL_Connection/_connection = connection
-	var/datum/BSQL_Operation/op = connectOperation
-	UNTIL(QDELETED(_connection) || op.IsComplete())
-	return !QDELETED(connection) && !op.GetError()
+	rustg_sql_disconnect_pool()
```
Comment (steamp0rt) [inline code/controllers/subsystem/dbcore.dm:125]:
```suggestion
/datum/controller/subsystem/dbcore/proc/IsConnected()
	if(!CONFIG_GET(flag/sql_enabled))
		return FALSE
	var/result = json_decode(rustg_sql_connected())
	world.log << result
	return result["status"] == "online"


```

**Comment by steamp0rt** [comment]
> Come on guys. Come on. Why are the only people who seem to care about fixing this known issue me, oranges, and mrty?

lmao

File: code/controllers/subsystem/dbcore.dm (hunk 3 above)
Comment (optimumtact) [reply code/controllers/subsystem/dbcore.dm:125]: please surface the errors to the sql log goof

File: code/controllers/subsystem/dbcore.dm [hunk 4]
```diff
@@ -62,68 +51,66 @@ SUBSYSTEM_DEF(dbcore)
...
-			if(query_db_version.NextRow())
-				db_major = text2num(query_db_version.item[1])
-				db_minor = text2num(query_db_version.item[2])
-				if(db_major != DB_MAJOR_VERSION || db_minor != DB_MINOR_VERSION)
-					schema_mismatch = 1 // flag admin message about mismatch
-					log_sql("Database schema ([db_major].[db_minor]) doesn't match the latest schema version ([DB_MAJOR_VERSION].[DB_MINOR_VERSION]), this may lead to undefined behaviour or errors")
-			else
-				schema_mismatch = 2 //flag admin message about no schema version
-				log_sql("Could not get schema version from database")
+			UNTIL(!query_db_version.in_progress)
```
Comment (optimumtact) [inline code/controllers/subsystem/dbcore.dm:100]: this one is not async on purpose I'm pretty sure

Comment (Iamgoofball) [reply code/controllers/subsystem/dbcore.dm:100]: I'm using it as my test query right now.

Comment (Iamgoofball) [reply code/controllers/subsystem/dbcore.dm:100]: I'm using it as my test query right now.

Comment (Iamgoofball) [reply code/controllers/subsystem/dbcore.dm:100]: It also was async, Execute() defaults to async with no args

Comment (optimumtact) [reply code/controllers/subsystem/dbcore.dm:100]: oh, okay then

Comment (optimumtact) [reply code/controllers/subsystem/dbcore.dm:100]: can you put the LOOP inside the execute command then please, not outside of it.

File: code/controllers/subsystem/dbcore.dm [hunk 5]
```diff
@@ -62,68 +51,66 @@ SUBSYSTEM_DEF(dbcore)
...
 		return FALSE
-
+	if(SSdbcore.IsConnected())
+		return TRUE
 	var/user = CONFIG_GET(string/feedback_login)
 	var/pass = CONFIG_GET(string/feedback_password)
 	var/db = CONFIG_GET(string/feedback_database)
 	var/address = CONFIG_GET(string/address)
 	var/port = CONFIG_GET(number/port)
+	var/list/connection_result = json_decode(rustg_sql_connect_pool(address, port, user, pass, db, "", "", ""))
```
Comment (steamp0rt) [inline code/controllers/subsystem/dbcore.dm:63]: golfball you're a fucking idiot you should be passing the BSQL config like max_connections and timeout to this

**Review by ghost — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/dbcore.dm (hunk 5 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/dbcore.dm:63]: we don't use BSQL anymore dude

Comment (optimumtact) [reply code/controllers/subsystem/dbcore.dm:63]: ^ actual state of goof

Comment (steamp0rt) [reply code/controllers/subsystem/dbcore.dm:63]: Tomorrow:

**Pull request closed without merging**